ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResource.java

```diff
@@ -404,6 +510,201 @@ private EndpointResponse handlePullQuery(
+        }
+
+        @Override
+        public Consumer<byte[], byte[]> getGlobalConsumer(Map<String, Object> map) {
+          return delegateKafkaClientSupplier.getGlobalConsumer(map);
+        }
+
+        @Override
+        public Admin getAdmin(Map<String, Object> config) {
+          return delegateKafkaClientSupplier.getAdmin(config);
+        }
+      };
+    }
+
+    @Override
+    public SchemaRegistryClient getSchemaRegistryClient() {
+      return delegate.getSchemaRegistryClient();
+    }
+
+    @Override
+    public Supplier<SchemaRegistryClient> getSchemaRegistryClientFactory() {
+      return delegate.getSchemaRegistryClientFactory();
+    }
+
+    @Override
+    public ConnectClient getConnectClient() {
+      return delegate.getConnectClient();
+    }
+
+    @Override
+    public SimpleKsqlClient getKsqlClient() {
+      return delegate.getKsqlClient();
+    }
+
+    @Override
+    public KafkaConsumerGroupClient getConsumerGroupClient() {
+      return delegate.getConsumerGroupClient();
+    }
+
+    @Override
+    public void close() {
+      delegate.close();
+    }
+
+    public boolean passed(Map<TopicPartition, Long> endOffsets) {
+      final Set<TopicPartition> finishedPartitions = new HashSet<>();
+      for (SubscriptionState subscriptionState : capturedConsumers) {
+        for (Map.Entry<TopicPartition, Long> entry : endOffsets.entrySet()) {
+          final TopicPartition topicPartition = entry.getKey();
+          try {
+            if (subscriptionState.isAssigned(topicPartition)) {
+              final SubscriptionState.FetchPosition position = subscriptionState.position(topicPartition);
+              if (position.offset == entry.getValue()) {
+                finishedPartitions.add(topicPartition);
+              }
+            }
+          } catch (RuntimeException e) {
+            e.printStackTrace();
+            throw e;
+          }
+        }
+      }
+      return finishedPartitions.equals(endOffsets.keySet());
+    }
+  }
+
+  private EndpointResponse handleStreamPullQuery(
+      final ServiceContext serviceContext,
+      final PreparedStatement<Query> statement,
+      final Map<String, Object> streamsProperties,
+      final CompletableFuture<Void> connectionClosedFuture,
+      final Map<TopicPartition, Long> endOffsets) {
+
+    // stream pull queries always start from earliest.
+    streamsProperties.put("auto.offset.reset", "earliest");
+
+    final ConfiguredStatement<Query> configured = ConfiguredStatement
+        .of(statement, SessionConfig.of(ksqlConfig, streamsProperties));
+
+    if (QueryCapacityUtil.exceedsPushQueryCapacity(ksqlEngine, ksqlRestConfig)) {
+      QueryCapacityUtil.throwTooManyActivePushQueriesException(
+              ksqlEngine,
+              ksqlRestConfig,
+              statement.getStatementText()
+      );
+    }
+
+    final ServiceContextOverride override = new ServiceContextOverride(serviceContext);
+
+    final TransientQueryMetadata query = ksqlEngine
```

Comment:
I guess this approach would allow us to do aggregates and all those other fun things as well, right?

Reply:
Yes, but the we should give some thought to how to surface amendments to users.

Reply:
Ah, looks like it will take more work to add it to the parser:
```
ksql> SELECT * FROM riderLocations WINDOW TUMBLING (SIZE 1 SECOND) GROUP BY PROFILEID;
Pull queries don't support WINDOW clauses. See https://cnfl.io/queries for more info.
Add EMIT CHANGES if you intended to issue a push query.
Caused by: Pull queries don't support WINDOW clauses.
ksql> SELECT * FROM riderLocations GROUP BY PROFILEID;
Pull queries don't support GROUP BY clauses. See https://cnfl.io/queries for more info.
Add EMIT CHANGES if you intended to issue a push query.
Caused by: Pull queries don't support GROUP BY clauses.

```

Probably for the best, since it will let us think about the ramifications.

Reply:
Thanks for checking~